[SDESK-7723] Fix add to planning item remaining locked

eos87 opened the pull request:
### Purpose
To fix the issue of the add to planning item remaining locked after closing the modal or after adding the item to coverage. In this PR we make to unlock the item if it was locked by the "Add to planning" modal in the first place.

### What has changed
- Refactored the `AddToPlanningController` to move unlock logic into a dedicated method and ensure it is called when the modal is closed
- Updated `ModalsContainer` to support an (optional) `onModalHide` callback 
- Adjusted event handling in `index.ts` to delay DOM cleanup to avoid race conditions with certain actions 
  
### Steps to test
Please follow the steps in the ticket comment https://sofab.atlassian.net/browse/SDESK-7723?focusedCommentId=141132

### Front-end checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] This pull request is adding missing TypeScript types to modified code segments where it's easy to do so with confidence
- [ ] This pull request is using TypeScript interfaces instead of prop-types and updates usages where it's quick to do so
- [ ] This pull request is using `memo` or `PureComponent` to define new React components (and updates existing usages in modified code segments)
- [ ] This pull request is replacing `lodash.get` with optional chaining and nullish coalescing for modified code segments
- [ ] This pull request is not importing anything from client-core directly (use `superdeskApi`)
- [ ] This pull request is importing UI components from `superdesk-ui-framework` and `superdeskApi` when possible instead of using ones defined in this repository.
- [ ] This pull request is not using `planningApi` where it is possible to use `superdeskApi`
- [ ] This pull request is not adding redux based modals
- [ ] In this pull request, properties of redux state are not being passed as props to components; instead, we connect it to the component that needs them. Except components where using a react key is required - do not connect those due to performance reasons.
- [ ] This pull request is not adding redux actions that do not modify state (e.g. only calling angular services; those should be moved to `planningApi`)

Resolves: [SDESK-7723]

Thanks for checking!



[SDESK-7723]: https://sofab.atlassian.net/browse/SDESK-7723?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eos87 committed:
Fix add to planning item remaining locked

SDESK-7723

thecalcc commented on index.ts:
```diff
@@ -68,16 +68,20 @@ window.addEventListener('planning:fulfilassignment', (event: CustomEvent) => {
 
-    // unmount the component and remove the element when the modal is closed
+    // clean up: unmount the React component and remove the DOM element after the modal closes.
+    // delay is added to ensure any pending actions complete before removal.
```
do you use `timeout` instead of `setTimeout` because it also handles UI updates? if `scope.$apply()` isn't needed here, it's better to remove angular wherever possible

Could `resolve` function be called after attempting to unlock locked items? That way we can avoid a timeout all together, which would also fail if action takes longer than 500ms

eos87 replied:
I found a way to avoid it :) Thanks for the suggestion.

thecalcc replied:
nice one! :-)

thecalcc commented on client/controllers/AddToPlanningController.tsx:
```diff
@@ -170,18 +170,24 @@ export class AddToPlanningController {
         }
     }
 
+    /**
```
nitpick: comment could be improved by saying something like: "Unlocks all items locked due to `add_to_planning` action or something like that

thecalcc commented on client/controllers/AddToPlanningController.tsx:
```diff
@@ -170,18 +170,24 @@ export class AddToPlanningController {
+    unlockAddToPlanningItem = () => {
+        const {lock_session, lock_action} = this.newsItem || {};
+
+        if (lock_session && lock_action === 'add_to_planning') {
```
existing constant can be used for `add_to_planning` from `PLANNING.ITEM_ACTIONS.ADD_TO_PLANNING.lock_action`

eos87 committed:
Refactor add to planning lock action handling

Replaces hardcoded 'add_to_planning' lock action strings with a constant from PLANNING.ITEM_ACTIONS. Simplifies modal cleanup logic in index.ts by removing unnecessary timeout delay and directly unmounting the React component.

SDESK-7723
